[Sema diagnostics & Tests] Add Sema diagnostics for detecting large integer literals that overflow Builtin.Int2048.

ravikandhadai opened the pull request:
and tests for checking the correctness of the diagnostics.

rdar://problem/39873993

ravikandhadai commented:
@DougGregor  Hi Doug, I have tried to implement the fix for detecting large integer overflows in the type checking phase, and added some tests including those that tests the raw value in the Enums. Let me know you what you think.  (ref: https://bugs.swift.org/browse/SR-7579?jql=text%20~%20%22Large%20integer%20literals%22)

ravikandhadai commented:
@swift-ci please smoke test

ravikandhadai commented:
@swift-ci please smoke test

huonw commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2446,6 +2446,10 @@ ERROR(no_MaxBuiltinIntegerType_found,none,
 ERROR(no_MaxBuiltinFloatType_found,none,
    "standard library error: _MaxBuiltinFloatType is not properly defined", ())
+ERROR(integer_literal_overflows_maxwidth, none,
+   "Integer literal size limit exceeded: '%2' needs %1 bits, "
```
Minor thing: diagnostics usually don't start with upper-case, the compiler itself is all lower-case and then Xcode will do the upper casing if necessary.

ravikandhadai commented:
@swift-ci please smoke test

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2446,6 +2446,9 @@ ERROR(no_MaxBuiltinIntegerType_found,none,
    "standard library error: _MaxBuiltinFloatType is not properly defined", ())
+ERROR(integer_literal_overflows_maxwidth, none,
+   "integer literal needs %1 bits, exceeding limit of %0 bits",
+   (unsigned, unsigned, StringRef))
```
Based on the use of this diagnostic the third parameter is supposed to be the number itself, but it's not used?

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -1950,6 +1950,33 @@ namespace {
+          litStr += intLitExpr->getDigitsText();
+          tc.diagnose(expr->getLoc(), diag::integer_literal_overflows_maxwidth,
+                      maxWidth, signedLitWidth, litStr);
+          return nullptr;
```
We are trying to reduce the number of `return nullptr` from CSApply because it means that constraint system shouldn't have formed solution like that at the first place. It seems that this check would be a better fit for `CSGen#visitLiteralExpr` because it uses raw data from the expression anyway, WDYT?

ravikandhadai replied:
@xdein Thanks for the comments. I would like to add it CSGen. However,  'MaxIntegerTypeDecl'  seems to be materialized only in CSApply just before the point where I have added this code (line 1933). I am wondering if it possible to know the maximum bit width of an integer literal in CSGen#visitLiteralExpr.

xedin replied:
@ravikandhadai It doesn't seem like it should be a problem to move the logic from 1933 - 1951 over to CSGen from here because it just does lookup into stdlib? It looks like nothing else in CSApply is using `MaxIntegerTypeDecl` anyway.

ravikandhadai replied:
@CodaFi It is possible but doesn't seem ideal. In the SIL generated by SILGen phase, integer literals get truncated to 2048 bits. So we don't see the original literal text unless we walk through the AST.  This also delays diagnostics (we don't need constant folding machinery to do this check.) Also, I guess handling Enum cases at the level of SIL opts may result in unnecessary overheads.

CodaFi replied:
You can walk back along Source Locs to AST nodes.  The literal instructions should support that.  @xedin is right here: failing the solution set this late for this kind of diagnostic isn’t great.

ravikandhadai replied:
@codaFi  I will try adding it to CSGen as suggested by @xedin.

xedin reviewed (changes requested):
Overall this is great, but I think it has to migrate to constraint generator instead...

ravikandhadai commented:
@swift-ci please smoke test

ravikandhadai commented:
@xedin I have moved the checks to CSGen phase as we had discussed. Let me know if it looks good.

xedin reviewed (approved):
LGTM! Please run these changes through `clang-format` before merging PR.

ravikandhadai committed:
[Sema] Add Sema checks for detecting large integer literals that
overflow Builtin.Int2048 and add tests for checking the
correctness of the diagnostics.

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci please test and merge
